maro/simulator/scenarios/vm_scheduling/business_engine.py

```diff
@@ -0,0 +1,482 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import os
+import shutil
+import tarfile
+from typing import Dict, List
+
+from yaml import safe_load
+
+from maro.backends.frame import FrameBase, SnapshotList
+from maro.cli.data_pipeline.utils import download_file, StaticParameter
+from maro.data_lib import BinaryReader
+from maro.event_buffer import CascadeEvent, EventBuffer, MaroEvents
+from maro.simulator.scenarios.abs_business_engine import AbsBusinessEngine
+from maro.simulator.scenarios.helpers import DocableDict
+from maro.utils.logger import CliLogger
+from maro.utils.utils import DottableDict
+
+from .common import DecisionPayload, Latency, PlaceAction, PostponeAction, PostponeType, VmRequestPayload
+from .cpu_reader import CpuReader
+from .events import Events
+from .frame_builder import build_frame
+from .physical_machine import PhysicalMachine
+from .virtual_machine import VirtualMachine
+
+metrics_desc = """
+
+total_vm_requests (int): Total VM requests.
+total_energy_consumption (float): Accumulative total PM energy consumption.
+successful_placement (int): Accumulative successful VM placement until now.
+successful_completion (int): Accumulative successful completion of tasks.
+failed_placement (int): Accumulative failed VM placement until now.
+total_latency (int): Accumulative used buffer time until now.
+total_subscriptions (int): Accumulative over-subscriptions.
```

Comment:
you mean ``over-load``?

Reply:
Should we need both metrices? over-load and over-subscription?

Reply:
then what's the meaning of over-subscription?

Reply:
Required resource is more than capacity -> over-subscription
Utilization resource is more than capacity -> overload

The meaning of over-subscription is that we could know the rate of the overload? Like there would be a slightly different if we have 100 over-subscription machine but only 1 overload happens vs 1 over-subscription machine with 1 overload?

Reply:
Since the VM is dynamically created and deleted, I'm not sure how we can define the number of over-subscription. In this case:

- PM.live_vm: [a, b, c]  <= not over-subscripted yet
- VM d is created, and PM.live_vm: [a, b, c, d] <= over-subscripted
- VM a is deleted, and PM.live_vm: [b, c, d] <= not over-subscripted now
- VM e is created, and PM.live_vm: [b, c, d, e] <= over-subscripted


As for the over-load, I think we can just count the number of VMs that are influenced when it happens (or the PM*ticks that over-load happens??), like:

- PM.live_vm: [a, b, c]  <= not over-subscripted yet
- VM d is created, and PM.live_vm: [a, b, c, d] <= over-subscripted
- VM a is deleted, and PM.live_vm: [b, c, d] <= not over-subscripted now
- VM e is created, and PM.live_vm: [b, c, d, e] <= over-subscripted
- **over-load happens, number of over-load += 4 ? (or number of over-load += 1*1)**
- **over-load still exists, number of over-load += 0 since these VM are all counted in? (or number of over-load += 1*1 again)**